rfc/JENKINS-41334-parallel-stage-execution.md

```diff
@@ -0,0 +1,76 @@
+## RFC: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334) - Parallel stage execution
+
+* Author: Andrew Bayer
+* Created Date: February 27, 2017
+* Approved Date: n/a
+* Target Release: 1.2
+* JIRA: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334)
+* Status: Draft
+
+### Problem Statement
+
+The `parallel` step does not fit smoothly into Declarative. It doesn't
+allow for parallel stage execution, it doesn't utilize the `agent`
+definition, etc. There needs to be the ability to run `stage`s in
+parallel, with the full power of the Declarative syntax enabled -
+i.e., `agent`, `post`, `when`, etc.
+
+### Syntax
+
+#### Proposed example
+
+```groovy
+stage('foo') {
+    parallel {
+        stage('first') {
+            steps {
+                echo "First branch"
+            }
+        }
+        stage('second') {
+            steps {
+                echo "Second branch"
+            }
+        }
+    }
+}
+```
+
+#### Details
+
+* A `stage` has to contain one and only one of `steps` or
+    `parallel`. They cannot be combined in the same `stage`.
+* `stage` configuration can be done on both a `stage` containing
+    `parallel`, or `stage`s within `parallel`, i.e., `agent`, `when`,
+    `post`, etc.
+* `agent`, `environment`, and `tools` specified on the "parent"
+    `stage` will apply to child `stage`s in the same way that top-level
```

Comment:
Perhaps.  I think we may want to run take a closer look specifically at the options allowed and when inheritance does/does not make sense, and where it might make sense to use something other than inheritance, potentially (leaving that one open here for now).